### PR TITLE
 fix <default> upgrade & improve backwards compatibility

### PR DESCRIPTION
**Pull request by tszirr**
This brings minor fixes for compatibility with previous scene file versions: It prevents the auto camelCase conversion from breaking \<default\> names, mangles names that collide with reserved IDs, and fixes 'diffuseReflectance' for diffuse BSDFs.

### CODE REVIEW

File: src/libcore/xml.cpp [hunk 1]
```diff
@@ -366,6 +369,20 @@ void upgrade_tree(XMLSource &src, pugi::xml_node &node, const Version &version)
...
+        // automatically rename reserved identifiers
+        for (pugi::xpath_node result: node.select_nodes("//@id")) {
+            pugi::xml_attribute id_attrib = result.attribute();
+            char const* val = id_attrib.value();
+            if (val && val[0] == '_') {
+                std::string new_id = std::string("ID") + val + "__UPGR";
+                Log(Warn, "Changing identifier: \"%s\" -> \"%s\"", val, new_id.c_str());
+                id_attrib = new_id.c_str();
+            }
+        }
```
Comment (Speierers) [inline src/libcore/xml.cpp:381]: Could you illustrate this by an example maybe? I am not sure to see

**Comment by tszirr** [comment]
Excerpt from the pool scene: https://cgg.mff.cuni.cz/~jirka/papers/2014/olpm/index.htm

```
	<bsdf type="phong" id="_01-Default">
		<rgb name="diffuseReflectance" value="0.67451 0.67451 0.67451"/>
		<rgb name="specularReflectance" value="0.0128 0.0128 0.0128"/>
		<float name="exponent" value="4"/>
	</bsdf>
```

With the patch, it is allowed to load, giving:
```
2020-04-16 15:34:39 WARN  main  [xml.cpp:387] Changing identifier: "_02-Default" -> "ID_02-Default__UPGR"
2020-04-16 15:34:39 WARN  main  [xml.cpp:387] Changing identifier: "_01-Default" -> "ID_01-Default__UPGR"
2020-04-16 15:34:39 WARN  main  [xml.cpp:387] Changing identifier: "_02-Default" -> "ID_02-Default__UPGR"
2020-04-16 15:34:39 WARN  main  [xml.cpp:387] Changing identifier: "_01-Default" -> "ID_01-Default__UPGR"
2020-04-16 15:34:39 WARN  main  [xml.cpp:387] Changing identifier: "_01-Default" -> "ID_01-Default__UPGR"
```

Without the patch, it crashes:
```
// src/libcore/xml.cpp
            if (string::starts_with(id, "_"))
                src.throw_error(
                    node, "invalid id \"%s\" in element \"%s\": leading "
                          "underscores are reserved for internal identifiers.",
                    id, node.name());
```

File: src/libcore/xml.cpp (hunk 1 above)
Comment (tszirr) [reply src/libcore/xml.cpp:381]: Excerpt from the pool scene: https://cgg.mff.cuni.cz/~jirka/papers/2014/olpm/index.htm

```
	<bsdf type="phong" id="_01-Default">
		<rgb name="diffuseReflectance" value="0.67451 0.67451 0.67451"/>
		<rgb name="specularReflectance" value="0.0128 0.0128 0.0128"/>
		<float name="exponent" value="4"/>
	</bsdf>
```

With the patch, it is allowed to load, giving:
```
2020-04-16 15:34:39 WARN  main  [xml.cpp:387] Changing identifier: "_02-Default" -> "ID_02-Default__UPGR"
2020-04-16 15:34:39 WARN  main  [xml.cpp:387] Changing identifier: "_01-Default" -> "ID_01-Default__UPGR"
2020-04-16 15:34:39 WARN  main  [xml.cpp:387] Changing identifier: "_02-Default" -> "ID_02-Default__UPGR"
2020-04-16 15:34:39 WARN  main  [xml.cpp:387] Changing identifier: "_01-Default" -> "ID_01-Default__UPGR"
2020-04-16 15:34:39 WARN  main  [xml.cpp:387] Changing identifier: "_01-Default" -> "ID_01-Default__UPGR"
```

Without the patch, it crashes:
```
            // src/libcore/xml.cpp
            if (string::starts_with(id, "_"))
                src.throw_error(
                    node, "invalid id \"%s\" in element \"%s\": leading "
                          "underscores are reserved for internal identifiers.",
                    id, node.name());
```

Comment (Speierers) [reply src/libcore/xml.cpp:381]: Thanks for your explanation, makes more sense now.

**Comment by Speierers** [comment]
Thanks for this PR @tszirr, it looks good to me, I will merge it now.

**Comment by merlinND** [comment]
Thank you for these PRs @tszirr!
I think that you mentioned testing this on a collection of Mitsuba 1 scenes. Would you mind sharing these, if it's not too much trouble?

**Comment by tszirr** [comment]
> Thank you for these PRs @tszirr!
> I think that you mentioned testing this on a collection of Mitsuba 1 scenes. Would you mind sharing these, if it's not too much trouble?

See invite. Tested most/all(?) of these, one scene has an invalid reflectance which needs to be fixed before loading (see error message on load). The living room scene fails to load because there is currently no alpha channel extraction feature, so in order to load remove the corresponding BSDF mask.

**Comment by merlinND** [comment]
Got the invite, thanks a lot! It's a useful resource.

**Pull request merged**